### PR TITLE
feat(shacl): added shacl lib

### PR DESCRIPTION
**Pull request by WilliamChelman**

### CODE REVIEW

File: libs/shacl/src/lib/services/shacl-data-graph-helper.service.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      });
+      return acc;
+    }, [] as ShNodeShape[]));
+  }
+
+  private getNodeShapesForShNode(nodeUri: string): ShNodeShape[] {
+    if (!this.shaclShapesGraphHelper.hasShNode) return [];
+    if (this.shNodeByNodeUri[nodeUri]) return this.shNodeByNodeUri[nodeUri];
+
+    const expectedNodeKind = isBlankNodeUri(nodeUri) ? SH.BlankNode : SH.IRI;
```
Comment (farah111) [inline libs/shacl/src/lib/services/shacl-data-graph-helper.service.ts:174]: we can use `getConcreteNodeKindOfValue`

**Review by farah111 — APPROVED** [review]

File: libs/shacl/src/lib/services/shacl-data-graph-helper.service.ts (hunk 1 above)
Comment (WilliamChelman) [reply libs/shacl/src/lib/services/shacl-data-graph-helper.service.ts:174]: `getConcreteNodeKindOfValue` only works with `JsonLdValue`, not strings ^^'

File: libs/shacl/src/lib/utils/get-path.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import { ExpandedJsonLdContainer, getOneValue, isBlankNodeUri } from '@cognizone/json-ld/core';
+import { SH } from '@cognizone/lod';
+
+import { PathMode, ShInversePathNode, ShPropertyShape } from '../models';
+
+// move to new ShaclHelper?
+export function getPath(propertyShapeUri: string, graph: ExpandedJsonLdContainer): PropertyPath {
```
Comment (srikanthv02) [inline libs/shacl/src/lib/utils/get-path.ts:7]: maybe rename it to `getPathMode` (looking at what is being returned?) 🤔

File: libs/shacl/src/lib/utils/get-path.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  if (path && mode) return { path, mode };
+  return { mode: 'unknown' };
+}
+
+function isInversePathNode(value: unknown): value is ShInversePathNode {
+  return typeof value === 'object' && value != null && (value as ShInversePathNode)[SH.inversePath] != null;
+}
+
+export type PropertyPath =
```
Comment (srikanthv02) [inline libs/shacl/src/lib/utils/get-path.ts:34]: and even this `PropertyPathMode` ? 🤔

File: libs/shacl/src/lib/utils/get-path.ts (hunk 2 above)
Comment (srikanthv02) [reply libs/shacl/src/lib/utils/get-path.ts:7]: also, even rename the file 🤔  (but you know it better :)  )

File: libs/shacl/src/lib/services/shacl-shapes-graph-helper.service.ts [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+      !!propertyShape?.[HANAMI.selection] ||
+      getConcreteNodeKindsOfPropertyShape(propertyShape, this.shapesGraph).includes(SH.BlankNode) ||
+      !!this.getConsolidatedPropertyShapesSortedBySpecificity(propertyShape['@id']).some(subProperty => {
+        if (subProperty['@id'] === propertyShape['@id']) return false;
+        return this.isPropertyReference(subProperty);
+      })
+    );
+  }
+
+  isSingle(nodeShapeUri: string, propertyPath: string): boolean {
```
Comment (srikanthv02) [inline libs/shacl/src/lib/services/shacl-shapes-graph-helper.service.ts:104]: we already have this defined in the [shacl-data-graph-helper.service.ts](https://github.com/cognizone/ng-cognizone/pull/63/files#diff-72f0265abd9a6a5085d564f94c95c131e91c224e621f3d652a7f8db406608e22) 🤔  (do we need to have these methods here again or can they be removed?)

**Review by srikanthv02 — CHANGES_REQUESTED** [review]

File: libs/shacl/src/lib/services/shacl-shapes-graph-helper.service.ts (hunk 4 above)
Comment (WilliamChelman) [reply libs/shacl/src/lib/services/shacl-shapes-graph-helper.service.ts:104]: I know it looks A LOT like being the same method, but one uses a `nodeUri`, the other a `nodeShapeUri`, which are 2 different things ^^'

**Pull request merged**